File: server/rag/retrieve/keyword_matcher.py
"""实体关键词匹配（领域纯函数，移植自 src/domain/entity/keywordMatcher.ts）。

- extract_matching_keywords：贪心最大匹配（长词优先、大小写不敏感）+ 全局包含匹配
- decompose_entity：未知实体分解为已知实体组合
字典（关键字列表）由调用方注入。
"""

from __future__ import annotations


def extract_matching_keywords(query: str, keywords: list[str]) -> list[str]:
    # dict 键保插入序：对齐 TS Set 的插入序（Python set 哈希序随机，
    # 最终长度降序稳定排序的并列项必须保持发现顺序）
    matched: dict[str, None] = {}
    query_lower = query.lower()
    sorted_kws = sorted(keywords, key=lambda kw: -len(kw))

    # 贪心最大匹配（从当前位置开始，按关键字长度降序尝试）
    i = 0
    qlen = len(query)
    while i < qlen:
        found = False
        for kw in sorted_kws:
            kw_len = len(kw)
            if i + kw_len > qlen:
                continue
            if query_lower[i : i + kw_len] == kw.lower():
                matched[kw] = None
                i += kw_len
                found = True
                break
        if not found:
            i += 1

    # 全局包含匹配（处理非连续出现）
    for kw in sorted_kws:
        if kw.lower() in query_lower:
            matched[kw] = None

    return sorted(matched, key=lambda kw: -len(kw))


def decompose_entity(unknown: str, known_names: list[str]) -> list[str]:
    """未知实体 → 所有为其子串（长度≥2）的已知实体，长度降序。"""
    results: list[str] = []
    unknown_lower = unknown.lower()

    for known in known_names:
        if len(known) >= 2 and known.lower() in unknown_lower:
            if not any(known in r for r in results):
                results.append(known)

    return sorted(results, key=lambda kw: -len(kw))

File: src/domain/entity/keywordMatcher.ts
// ============================================================
// 实体关键词匹配（领域规则，纯函数，零基础设施依赖）
// 从 entityRouter / queryRewriter 拆出的匹配算法：
//   - extractMatchingKeywords：贪心最大匹配 + 全局包含匹配
//   - decomposeEntity：未知实体分解为已知实体组合
// 字典（关键字列表）由调用方注入（来自 EntityRepository）
// ============================================================

/**
 * 从用户问题中提取匹配的实体关键字
 * 使用贪婪最大匹配（优先长词，大小写不敏感），再全局包含匹配（处理非连续出现）
 *
 * @param query - 用户查询
 * @param keywords - 已知实体关键字字典（任意顺序，内部按长度降序处理）
 */
export function extractMatchingKeywords(query: string, keywords: string[]): string[] {
  const matched: Set<string> = new Set();
  const queryLower = query.toLowerCase();
  const sorted = [...keywords].sort((a, b) => b.length - a.length);

  // 贪心最大匹配（从当前位置开始，按关键字长度降序尝试）
  let i = 0;
  while (i < query.length) {
    let found = false;

    for (const kw of sorted) {
      const kwLen = kw.length;
      if (i + kwLen > query.length) continue;
      // 大小写不敏感比较
      if (queryLower.slice(i, i + kwLen) === kw.toLowerCase()) {
        matched.add(kw);
        i += kwLen;
        found = true;
        break;
      }
    }

    if (!found) i++;
  }

  // 也尝试在整个 query 中搜索（处理非连续匹配的情况）
  for (const kw of sorted) {
    if (queryLower.includes(kw.toLowerCase())) {
      matched.add(kw);
    }
  }

  return [...matched].sort((a, b) => b.length - a.length);
}

/**
 * 实体分解：将未知实体分解为已知实体的组合
 * 策略：从已知实体列表中查找所有是未知实体子串的实体
 * 例如："南方电网供应链管理平台项目" → ["南方电网", "供应链管理平台"]
 */
export function decomposeEntity(unknown: string, knownNames: Iterable<string>): string[] {
  const results: string[] = [];
  const unknownLower = unknown.toLowerCase();

  for (const known of knownNames) {
    if (known.length >= 2 && unknownLower.includes(known.toLowerCase())) {
      if (!results.some(r => r.includes(known))) {
        results.push(known);
      }
    }
  }

  return results.sort((a, b) => b.length - a.length);
}
